### PR TITLE
feat(auth): Support username/passowrd and JWT auth

### PR DESCRIPTION
**Pull request by gaocegege**
This PR is to support auth and no-auth mode in envd-server.

With auth enabled, users need to register first, then use `envd login --username xx --password xx` to login.

With auth disabled, users can use `envd login` directly, to register and login with a random-generated UID.

Signed-off-by: Ce Gao <cegao@tensorchord.ai>

### CODE REVIEW

**Comment by gaocegege** [comment]
TODO: Support https://github.com/swaggo/swag/issues/144 in api doc

File: api/types/auth.go [hunk 1]
```diff
@@ -6,26 +6,24 @@ package types
 
 // AuthRequest contains authorization information for connecting to a envd server.
 type AuthRequest struct {
```
Comment (xieydd) [inline api/types/auth.go:8]: Maybe can use authn,separate from authz to avoid confusion.

File: pkg/service/user/user.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/jackc/pgx/v4"
+
+	"github.com/tensorchord/envd-server/pkg/query"
+)
+
+var (
+	// ExpirationTimeDefault is the default expiration time of the token.
+	expirationTimeDefault = time.Hour * 24 * 365
+	// jwtSecretDefault is the secret used to sign the token.
+	jwtSecretDefault = "envd-secret"
```
Comment (VoVAllen) [inline pkg/service/user/user.go:21]: I think the secret should be generated and should not be leaked. Should be the same for one server, but different for each server

File: pkg/util/sql/schema.sql [hunk 3]
```diff
@@ -1,10 +1,20 @@
 -- Users
 CREATE TABLE IF NOT EXISTS users (
   id BIGSERIAL PRIMARY KEY,
-  identity_token text NOT NULL,
+  login_name varchar(100) NOT NULL,
+  password_hash text NOT NULL,
```
Comment (VoVAllen) [inline pkg/util/sql/schema.sql:5]: is there any better type instead of text for this?

Comment (xieydd) [reply pkg/util/sql/schema.sql:5]: Text may be more convenient backwards compatibility, compared to fixed length type like varchar(n). I think text is ok.

File: pkg/service/user/user.go (hunk 2 above)
Comment (gaocegege) [reply pkg/service/user/user.go:21]: I do not think we should generate it. Maybe we can make it a config. WDYT

File: api/types/auth.go (hunk 1 above)
Comment (gaocegege) [reply api/types/auth.go:8]: LGTM

File: pkg/service/user/user.go (hunk 2 above)
Comment (VoVAllen) [reply pkg/service/user/user.go:21]: config is fine. I'm not sure whether we need to store this in db

File: test/.editorconfig [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# EditorConfig is awesome: https://EditorConfig.org
```
Comment (kemingy) [inline test/.editorconfig:1]: Same here

File: pkg/service/user/user.go (hunk 2 above)
Comment (kemingy) [reply pkg/service/user/user.go:21]: Both the JWT secret and expiration time should be configurable. The default one should be much more complex.

I think we don't need to store them in DB.

File: client/client.go [hunk 5]
```diff
@@ -23,6 +23,9 @@ var ErrRedirect = errors.New("unexpected redirect in response")
 // Client is the API client that performs all operations
 // against a docker server.
 type Client struct {
+	auth     bool
```
Comment (VoVAllen) [inline client/client.go:26]: Why does the client need to take care of whether it's auth or not?

Comment (VoVAllen) [reply client/client.go:26]: Also I think a random generated name like https://github.com/tensorchord/envd/blob/main/pkg/util/fileutil/namegenerator.go is better than random token as the username

Comment (VoVAllen) [reply client/client.go:26]: I changed my mind. I think it's better to force users to provide a username and password, to make it a consistent behavior across auth and no-auth. It won't make any difficult for users to choose a username for themselves

Comment (gaocegege) [reply client/client.go:26]: > I changed my mind. I think it's better to force users to provide a username and password, to make it a consistent behavior across auth and no-auth. It won't make any difficult for users to choose a username for themselves

client needs to know if the jwt token is needed.

Comment (gaocegege) [reply client/client.go:26]: > For noauth scenario, you can still sign a JWT token to the client. Thus it will make no difference at client side

SGTM

**Comment by gaocegege** [comment]
PTAL

**Review by VoVAllen — APPROVED** [review]
LGTM

**Review by kemingy — APPROVED** [review]

**Pull request merged**